Extract upload code in a shared helpers module.

lraucy opened the pull request:
### What and why?

Refactor `multipart/form-data` payload uploads in a single shared module to reduce code duplication.

### How?

- Introduce a new `helpers/upload` module with helpers to upload `multipart/form-data` payloads.
- Use this new module in `sourcemap upload`, `dsyms upload` and `commit upload`
- Logging & metrics reporting are still specific to each commands
- `junit upload` and `dependencies upload` are not using this helper yet but could be migrated afterwards.

### Review checklist

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration)

lraucy commented on src/commands/sourcemaps/upload.ts:
```diff
@@ -141,47 +137,34 @@ export class UploadCommand extends Command {
+          gitCommitSha: repositoryData.hash,
+          gitRepositoryPayload: repositoryPayload,
+          gitRepositoryURL: repositoryData.remote,
+        })
```
I introduced a new interface GitData to group all git-related information into a single object (the object is either present entirely or not, no real use case to have these fields partially defined).

lraucy commented on src/commands/sourcemaps/upload.ts:
```diff
@@ -141,47 +137,34 @@ export class UploadCommand extends Command {
-          sourcemapPath,
-          version: this.releaseVersion!,
-        }
+        return new Sourcemap(minifiedFilePath, minifiedURL, sourcemapPath)
```
`cliVersion`, `projectPath`, `service`, `version` are all common to all sourcemaps a single run, so no need to duplicate the info in all `Sourcemap` object.

buranmert reviewed (approved):
regarding `dsyms` folder, LVGTM 👍

m-rousse commented on src/helpers/upload.ts:
```diff
@@ -0,0 +1,129 @@
+}
+
+export interface UploadOptions {
+  /** ApiKeyValidator (optional) throws an InvalidConfiguratonException when upload fails because
```
```suggestion
  /** ApiKeyValidator (optional) throws an InvalidConfigurationException when upload fails because
```

m-rousse commented on src/helpers/upload.ts:
```diff
@@ -0,0 +1,129 @@
+
+  /** Callback to execute before retries
+   */
+  onRetry(e: Error, attempts: number): void
```
Nit: to be consistent with other callback definition
```suggestion
  onRetry(error: Error, attempts: number): void
```

m-rousse commented on src/helpers/apikey.ts:
```diff
@@ -1,28 +1,57 @@
+import {BufferedMetricsLogger} from 'datadog-metrics'
 
+import {InvalidConfigurationError} from './errors'
+import {} from './utils'
```
I'm not sure if our linter allows this, but you should be able to write this as `import './utils'`.
Also, why is it needed to import the file but no function from it?

lraucy replied:
Strange, that must be a leftover. I removed it!

jybp reviewed (approved):
Thanks for this refactoring, LGTM 👍

yoannmoinet commented on src/commands/commit/upload.ts:
```diff
@@ -30,32 +32,50 @@ export class UploadCommand extends Command {
+      const status = await this.uploadRepository(requestBuilder)(payload, {
+        apiKeyValidator,
+        onError: (e) => {
+          this.context.stdout.write(renderFailedUpload(e.message)), metricsLogger.logger.increment('failed', 1)
```
is this correct?
Shouldn't it be:
```suggestion
          this.context.stdout.write(renderFailedUpload(e.message))
          metricsLogger.logger.increment('failed', 1)
```

yoannmoinet reviewed (commented):
LGTM.

Just a weird formatting thing.
Also, shouldn't `src/helpers/upload.ts` be tested as well?

lraucy commented:
> Also, shouldn't `src/helpers/upload.ts` be tested as well?

Very good point. I added some tests [in this commit](https://github.com/DataDog/datadog-ci/pull/323/commits/bb286e649a8784e5381b0cdba1ff987ad0391115). I also changed the exported ApiKeyValidator into an interface to ease testing without doing weird type casting.

jybp commented on src/helpers/__tests__/upload.test.ts:
```diff
@@ -0,0 +1,193 @@
+      )
+      expect(result).rejects.toMatch('errorApiKey')
+      expect(mockCreate).toBeCalledTimes(1)
+      expect(uploadCallback).toBeCalledTimes(1)
```
I was about to suggest to add `expect(verifyKey).toBeCalledTimes(1)` but it doesn't seem to work, no idea why.

lraucy replied:
Yes I initially added this check but couldn't make it work, I wonder if this is related to how exception/rejected promise handling work with jest. I didn't dig further because I realized if the rejected promise matched `errorApiKey` that proves that the mocked `verifyKey` function has been called.
